[Form] fixed BC break with pre selection of choices with `ChoiceType` and its children

HeahDude opened the pull request:
| Q | A |
| --- | --- |
| Branch | 2.7+ |
| Bugfix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #18173, #14712, #17789 |
| License | MIT |
| Doc PR | - |
- f7eea72 reverts BC break introduced in #17760
- 58e8ed0 fixes pre selection of choice with model values such as `false`, `null` or empty string without BC break.

`ChoiceType` now always use `ChoiceToValueTransformer` or `ChoicesToValuesTransformer` depending on `multiple` option.
Hence `CheckboxListMapper` and `RadioListMapper` don't handle the transformation anymore.

HeahDude committed:
Revert "[Form] refactor `RadioListMapper::mapDataToForm()`"

closes #18173.

This reverts commit 8f918e5f84c53826d025cbb1f3a0bfe093acc4e0.

HeahDude committed:
[Form] refactor CheckboxListMapper and RadioListMapper

fixes #14712 and #17789.

`ChoiceType` now always use `ChoiceToValueTransformer` or
`ChoicesToValuesTransformer` depending on `multiple` option.
Hence `CheckboxListMapper` and `RadioListMapper` don’t handle
the transformation anymore.
Fixes pre selection of choice with model values such as `null`,
`false` or empty string.

HeahDude commented on src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php:
```diff
@@ -1247,7 +1279,7 @@ public function testSubmitSingleExpandedWithEmptyChild()
         $form->submit('');
 
-        $this->assertNull($form->getData());
+        $this->assertEmpty($form->getData());
```
This looks like a collateral fix to me :)

HeahDude replied:
@webmozart you need to read this test in particular, the others are just using an empty string instead of `null` as `viewValue` but this one concerns the data.

Also this PR fixes the failed tests added originally in #17760 (ref https://github.com/symfony/symfony/pull/17760/files#diff-533d45d8bb41e489890ad12df7a1a4b7R163).

HeahDude replied:
This time it just don't break the data transformers call :)

javiereguiluz commented:
@HeahDude thanks for this fix (and for everything you do for the Form component).

Just a minor comment: it seems that the `| Bug fix?      | yes/no` row disappeared from the PR description table (it is usually displayed after the `Branch?` row). This information is important for mergers. Could you please include it? Thanks!

HeahDude commented:
@javiereguiluz done :)

HeahDude commented on src/Symfony/Component/Form/Extension/Core/DataMapper/RadioListMapper.php:
```diff
@@ -26,23 +26,17 @@
-    public function mapDataToForms($data, $radios)
+    public function mapDataToForms($choice, $radios)
     {
+        if (!is_string($choice)) {
```
Should I test null of force a string casting before throwing here ?

HeahDude replied:
@webmozart I re-check the code and here the view data should always be a string.

In contrary, since `mapsFormsToData()` happens before transformation we should accept `null` there as well.

Do you agree or am I missing something ? Thanks.

HeahDude commented:
Failures seem unrelated.

nakashu commented:
:+1:

stof commented on src/Symfony/Component/Form/Extension/Core/DataMapper/CheckboxListMapper.php:
```diff
@@ -46,22 +35,12 @@ public function mapDataToForms($choices, $checkboxes)
         foreach ($checkboxes as $checkbox) {
             $value = $checkbox->getConfig()->getOption('value');
-            $checkbox->setData(isset($valueMap[$value]) ? true : false);
+            $checkbox->setData(in_array($value, $choices, true) ? true : false);
```
the ternary operator turning a boolean into the same boolean looks weird to me

HeahDude replied:
@stof I first felt the same, as you can see I did to match the previous style (in both checkbox and radio list mappers).

I guess it's a kind of explicit casting since this is where choices string values are converted to their boolean "selected" state values.

Thanks for the review.

dmaicher replied:
But technically 

```php
in_array($value, $choices, true) ? true : false
```

is the same as

```php
in_array($value, $choices, true)
```

or am I missing something? :)

fabpot replied:
I would indeed remove those unneeded ternary operators when the left side is already a bool.

HeahDude replied:
Fair enough then.

HeahDude replied:
Fixed.

HeahDude commented on src/Symfony/Component/Form/Extension/Core/DataMapper/RadioListMapper.php:
```diff
@@ -59,10 +57,9 @@ public function mapFormsToData($radios, &$choice)
+                $choice = $radio->getConfig()->getOption('value');
 
-                return;
+                return $choice;
```
Just realised this return statement is useless.

HeahDude commented on src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php:
```diff
@@ -61,8 +61,8 @@ public function buildForm(FormBuilderInterface $builder, array $options)
-                ? new CheckboxListMapper($options['choice_list'])
-                : new RadioListMapper($options['choice_list']));
+                ? new CheckboxListMapper()
+                : new RadioListMapper());
```
Little detail, it's now short enough to be inlined here imho, isn't it ?

fabpot replied:
It could have been inlined before as well :)

HeahDude replied:
done :)

HeahDude commented:
Comments addressed.

HeahDude commented:
`TwigBridge` dependency to `EventDispatcher` seems to be broken with php 7 build in travis.

xabbuh commented:
@HeahDude Can you check if rebasing fixes the tests?

HeahDude commented:
@xabbuh It did. Thanks!

webmozart commented on src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php:
```diff
@@ -133,30 +131,13 @@ public function buildForm(FormBuilderInterface $builder, array $options)
-            }
-        } elseif ($options['multiple']) {
-            // <select> tag with "multiple" option
+        if ($options['multiple']) {
```
I don't fully understand why this is necessary. It seems to me the transformer and the data mappers are doing the same thing already. Can you add some explanation?

HeahDude commented on src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php:
```diff
@@ -133,30 +131,13 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+        }
 
-            if (!$options['multiple']) {
-                // For radio lists, transform empty arrays to null
```
@webmozart it also prevent this hack

HeahDude commented on src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php:
```diff
@@ -1194,7 +1226,7 @@ public function testSubmitSingleExpandedNonRequiredFalse()
 
         $this->assertNull($form->getData());
-        $this->assertNull($form->getViewData());
+        $this->assertEmpty($form->getViewData());
```
@webmozart "fixing" those tests seems logic to me as the placeholder is an empty string.

And it is coherent with this [comment](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php#L329).

webmozart replied:
I see. You should rather use `assertSame('', ...)` though if you want to ensure this is an empty string. `assertEmpty()` accepts a lot of values (`null`, `false`, ...).

HeahDude replied:
Absolutely! Will do :)

HeahDude commented on src/Symfony/Component/Form/Extension/Core/DataMapper/CheckboxListMapper.php:
```diff
@@ -70,6 +49,10 @@ public function mapDataToForms($choices, $checkboxes)
      */
     public function mapFormsToData($checkboxes, &$choices)
     {
+        if (!is_array($choices)) {
```
Is this check needed ? Shouldn't we type hint here ?

HeahDude commented on src/Symfony/Component/Form/Extension/Core/DataMapper/CheckboxListMapper.php:
```diff
@@ -46,22 +35,12 @@ public function mapDataToForms($choices, $checkboxes)
-                $e->getCode(),
-                $e
-            );
+            throw new UnexpectedTypeException($choices, 'array');
```
same here

HeahDude commented:
Ok there was a reason why I use `assertEmpty`, it's because `$extraData` in as empty array while `$viewData` is an empty string.

I fixed the assertions accordingly.

Rebased and squashed! Needs a final review here. Thanks!

HeahDude commented:
There is a little failure in Doctrine bridge because it needs that PR to be merged first.

I've updated the dependency to `2.7.12` but travis does not seem to be aware of it yet.

Is it OK ?

xabbuh commented on src/Symfony/Bridge/Doctrine/composer.json:
```diff
@@ -22,7 +22,7 @@
         "symfony/stopwatch": "~2.2",
         "symfony/dependency-injection": "~2.2",
-        "symfony/form": "~2.7,>=2.7.1",
+        "symfony/form": "~2.7,>=2.7.12",
```
Can you try `~2.7.12|~2.8.5` please?

HeahDude commented:
@xabbuh Done, let's wait & see :)

HeahDude commented:
Nope :(

HeahDude commented:
@xabbuh Can this be related to the last releases messed tags ?

xabbuh commented:
@HeahDude This looks good now (the `deps=low` job is passing).

HeahDude commented:
Great!

dmaicher commented:
Any news here? Waiting for this fix so I can upgrade to Symfony 2.8.* :)

fabpot commented:
Thank you @HeahDude.

HeahDude commented:
@dmaicher, happy upgrade! :)
